### ISSUE TITLE
"The neighbourhood graph is not connected."

### ISSUE THREAD

**Issue by jejjohnson** [lisitsyn/tapkee#28]
#### Overview

Hello!

So I was trying to use your dimension reduction package in relation to hyperspectral imagery as you cited as an application within your benchmarks and paper. So I cloned your benchmarks repository and ran your script to perform the benchmark and I didn't run into any problems. However, I downloaded the library and wanted to mess around with dimension reduction for hyperspectral imagery and I get this glaring error no matter which similar hyperspectral data set I used - **[warning] The neighborhood graph is not connected.** This leads my results to having nan values. This is consistent between the tapkee CLI inputs as well as the shogun-toolbox inputs.

#### Specific Issue

So let's say I run this command in the CLI:

*./tapkee -i aviris.dat -o aviris_dimred.dat -m isomap -td 20 -k 25 --benchmark*

The method will run fine except for the  **[warning] The neighborhood graph is not connected.** error that pops up no matter what simple parameters I use, i.e. k-nearest neighbors, target dimension, eigensolver method. Upon closer inspection, I see that the output file is simply a list of *nan* for each column. 

So I tried to vary the dimension reduction techniques, i.e. pca, laplacian eigenmaps, neighborhood preserving projection, local linear embedding, etc, and I found that the eigendecomposition would always fail as an error would pop up saying *Some error occured: eigendecomposition failed*. Now I varied the knn, the dimension and the methods and I found that the only methods that would produce a solution with actual values in the output file were the stochastic methods; e.g. t-stochastic neighbourhood embedding and stochastic proximity embedding; and MDS method. Sometimes my solution would be nonsense - like if the input data resides between 0 and 1, the output data should as well and sometimes these algorithms didn't produce that - but that is to be expected with stochastic methods. And the MDS will put out some nan values as well but not all of them.

I also tried to use my own hyperspectral data set albeit it was much bigger; 145x145 with 200 dimensions. I created a flattened image of it so that the dimensions were 21025x200. However, I still got the same errors except it just took longer to process.

I even get the same issue if I try to use the Shogun-Toolbox to try and enter the same data sets via Python. Same error except it will not even produce a result and stops the algorithm altogether.

#### Question

So, what did you do for your graph that produced actual results? I looked into your script and I didn't find any extra commands that would allow you overcome this error. Maybe I am missing something in your script for using Tapkee on the aviris dataset? What commands should I enter or vary so that I can get a sensible solution with your dataset and package? Is there some sort of preprocessing step that you (or I could) do to avoid this issue?

Thank you for your time!

**Comment by lisitsyn** [comment]
@jejjohnson thanks for such detailed explanation of the error. 

There is a long missed feature of handling disconnected neighborhood graphs, sorry about that. I guess you're referring https://github.com/lisitsyn/tapkee_jmlr_benchmarks which has setting of `k=250` for the aviris data. Such a big number of neighbors was selected by us to mitigate disconnected graphs issue and to benchmark the library with high-dimensional high-k setting.

In your experiments you used `k=25` which leads to disconnected graphs so all neighborhood-based methods (isomap, lle, ltsa, ...) fail badly. I am not sure what's the best way to overcome this issue with low `k` values. It could take the largest connected component but it would return a subset of data then. It could also 'repair' graph - I like this solution, but didn't find enough time to implement it yet.

**Comment by jejjohnson** [comment]
@lisitsyn Thank you for your rapid response!

I think understand what you're saying. Sorry I wasn't clear on which toolbox I was referring to. I've actually never had the privilege of coming across a disconnected graph error in many toolboxes I have use. In the theory, we always assume the graph is connected but there are ways to mitigate this issue; like the 'repairing' method you mentioned; which many authors of papers kind of slide underneath the rug. To be honest, it's very hard to find working code from many scientists within the remote sensing community of how they did their dimension reduction methods so I have a hard time understanding/believing some of their work. 

In any case, I did vary my knn value to be a max of 500 but at that point my PC screamed at me to stop so I did. That being said, I still wasn't able to get a connected neighborhood graph between `k=100` to `k=500`. Hyperspectral datasets I use seem to be the only issue I've seen with this package. Every other application dataset I throw at it, it handles beautifully and very fast! So if anything, that's kind of motivating because it means it's a difficult problem to solve! =)

But thank you very much for this great toolbox and for also answering my questions! Best of luck to you!
